src/libopensc/asn1.c

```diff
@@ -807,10 +825,45 @@ sc_asn1_decode_object_id(const u8 *inbuf, size_t inlen, struct sc_object_id *id)
 	sc_init_oid(id);
 	octet = id->value;
 
-	a = *p;
-	*octet++ = a / 40;
-	*octet++ = a % 40;
-	inlen--;
+	/* The first octet can be 0, 1 or 2 and is derived from the first byte */
+	a = MIN(*p / 40, 2);
+	*octet++ = a;
+
+	/* The second octet fits here if the previous was 0 or 1 and second one is smaller than 40.
+	 * for the value 2 we can go up to 47. Otherwise the first bit needs to be set
+	 * and we continue reading further */
+	if ((*p & 0x80) == 0) {
+		*octet++ = *p - (a * 40);
+		inlen--;
+	} else {
+		/* Use unsigned type here so we can process the whole INT range */
+		unsigned int value = (*p & 0x7F);
+		inlen--;
+		while (inlen && *p & 0x80) {
+			/* Limit the OID values to int size and do not overflow */
+			if (value > (UINT_MAX>>7)) {
+				sc_init_oid(id);
+				return SC_ERROR_NOT_SUPPORTED;
+			}
+			p++;
+			value <<= 7;
+			value |= *p & 0x7F;
+			inlen--;
```

Comment:
this looks like code duplication with the below code

wouldn't it be simpler to decode all subidentifiers and then split the first one into two and insert new values to the array? less code, less special cases = less potential bugs

Reply:
True. I will simplify that.